### DISCUSSION TITLE
How to retrieve data from a specific row and column to another table through automation?

### DISCUSSION THREAD

**Discussion by Bergem93** [Budibase/budibase#9710] [Help] [answered]
Hi,

First of, i'm not at all a developer by any means, so if am missing something obvious, sorry in advance. 

**Purpose:**
Create a table (Service Monthly Cost table) which stores historic data about service costs per month (Service table), which can later be visualized through a bar chart. This table will include a column per service and a date column, and each row will contain that months monthly cost total and a date stamp. 
To get data over time, I want to automate the operation with a Cron trigger at the end of each month. The automation should pull Monthly Cost data for all Services in the "Services" table and write to "Service Monthly Cost" table.

**Problem**
I can't find the correct syntax to reference specific rows and columns in the "Service" table to retrieve Monthly Cost Total.  
I've tried muliple variation of $(Services.0.Monthly Cost Total); and the script runs but writes NULL to all rows. Also in the Test details, it says the columns JS didnt execute, which could be a reason for why the data = NULL, but in the db itself everything is working as expected. Any pointers?

        "Number of Users": "Error while executing JS",
        "Monthly Cost Total": "NaN",

**Comment by melohagan** [comment]
Hey @Bergem93 

This syntax is not correct: `$(Services.0.Monthly Cost Total);`. The syntax should be something like 

```js
return $("Services")?.[0]?.["Monthly Cost Total"];
```

If that doesn't work for you, then some screenshots would help!

**Reply by Bergem93** [reply]
Hi @melohagan,

I tried the supplied syntax and got an "Error while executing JS".

The script commented out is what i have tried with so far. It runs, but the row returns NULL. 
![Javascript relationship code](https://user-images.githubusercontent.com/5398191/219597555-81545c83-e512-40c9-94dc-f8ddc82da89b.png)

**Reply by melohagan** [reply]
Sorry I'm only noticing that this is for an automation. I think there is possibly a limitation with automations and accessing relational data. 
Could you send some screenshots of your automation with the bindings you are using and I can try to reproduce.

**Reply by Bergem93** [reply]
No problem!

Here is the current automation. Very simple, its set to trigger at a certain time at the end of each month.
Then create a new row in Service Monthly Cost with data from Services along with a datestamp.

This is the Automation. TripleTex JS contains the script you supplied, as well as my own which is commented out. Date is simply "new Date();" return, and is currently the only field that works as intended.
![Automation](https://user-images.githubusercontent.com/5398191/219645697-aae22781-3343-494c-989a-19d93cefc015.png)

This is the Services Monthly Cost table. The first row i've added manually to illustrate how i want it to look like.
![Services Monthly Cost total](https://user-images.githubusercontent.com/5398191/219645737-6c677e11-1a2d-49cb-a256-22e0b13ee9a2.png)

This is the Services table with data should be fetched from.
![Services Table](https://user-images.githubusercontent.com/5398191/219645758-3c952f0a-8d1a-4ca1-8715-e7108da1787f.png)

**Comment by melohagan** [comment, marked as answer]
Thanks for the screenshots @Bergem93 

I think you're missing a step in your automations. You'll need something like this:

**Cron Trigger ->**
1. Query Rows (In your case the Services table. You may want to add a filter for the month if needed?)
2. Create Row (using a binding to sum the costs from the query in the previous step)

For example:
**Expenses** table
![Screenshot 2023-02-17 at 13 29 16](https://user-images.githubusercontent.com/101575380/219665374-af57fef9-a70a-4030-94a9-2342ca7d93e8.png)

**Step 1**
![Screenshot 2023-02-17 at 13 29 52](https://user-images.githubusercontent.com/101575380/219665511-2a0ca288-faa9-4e3e-b2fc-b78a85a1a5bc.png)

**Step 2**
![Screenshot 2023-02-17 at 13 30 25](https://user-images.githubusercontent.com/101575380/219665637-ebb18200-14cb-4992-be8a-eaa3ce423c5a.png)

**Output**
![Screenshot 2023-02-17 at 13 30 55](https://user-images.githubusercontent.com/101575380/219665735-273fef9c-903a-496b-ba9c-01ae78dfed17.png)

**Reply by Bergem93** [reply]
Thank you so much for the help so far, we are so close....

I tried replicating your code, but it wont accept the Syntax. This code will also add all the rows together which i dont really need since that is already done in the Service Table. I'm only trying to fetch Monthly Cost Total. 
![image](https://user-images.githubusercontent.com/5398191/219684674-df57c873-473e-49b2-8e9b-b86a8dfcc570.png)

If I try to force it and just type the code directly into the field without entering the "editor", it just return the code snippet in the db.
![image](https://user-images.githubusercontent.com/5398191/219685317-d9afa866-89c9-4f39-9b9c-c1eace2de7ec.png)

However, by using only {{ steps.1.rows }} it printed the entire service rows to the column, so im only missing the correct handlebar to pinpoint Monthly Cost Total. I found this thread (https://github.com/Budibase/budibase/discussions/7772) which I think is the solution i need, but im failing to replicate it with my variables. 

Would the correct handlebar be something like `{{ steps.1.rows.0.Services.0.Monthly Cost total }}` ?
For the record, i've tried that syntax and it returned NULL.

**Reply by melohagan** [reply]
You just need a space I believe: `{{ sum (pluck steps.1.rows 'Monthly Cost Total') }}

Nice find with that discussion. In your case because the column has a space, I believe you'll need to wrap it in square brackets. Possibly along the lines of: `steps.1.rows[0][Services][0][Monthly Cost total]` or `steps.1.rows.0.Services.0.[Monthly Cost total]` maybe.

**Reply by Bergem93** [reply]
Hmm, i've tried many different variations now of the binding, but can't get it to work. It return an empty field. I've also tried to fetch the _id so that the only change from his binding to mine is Capo --> Services, but it still returns an empty field. 

I managed to get the pluck function to work, but that only gets the whole array and not related to the specific Service i want. 
Do you have any suggestions?

**Reply by Bergem93** [reply]
@melohagan Any possibility that you could advise me a bit further on this. Not been able to progress on this issue since last time.

**Reply by melohagan** [reply]
Had you added the Query Rows step with the filter to get the related data?

**Reply by Bergem93** [reply]
@melohagan Yes, im querying the relevant rows, then creating a new row with the binding {{ steps.1.rows.0.Services.0.[Monthly Cost Total] }}

Returns empty.

![Service monthly cost 1](https://user-images.githubusercontent.com/5398191/223464056-7f124eaa-af8c-44e8-8fc9-3b5faaa2497b.png)
![Service monthly cost 2](https://user-images.githubusercontent.com/5398191/223464079-2f57af3e-44e4-4bb3-8c71-2e5c0b324fb8.png)

**Reply by melohagan** [reply]
Hey @Bergem93 

If Step 1 is returning *Services* then Step 2 should look more like: `{{ steps.1.rows.0.[Monthly Cost total] }}`

If you're still having issues, I'd also try changing the field name to remove spaces to see if that helps. Then you could do `{{ steps.1.rows.0.MonthlyCostTotal }}`

Also make sure that you have **Use bindings** selected from the dropdown beside `Add looping`, although it appears to be in the binding mode alright.

**Reply by Bergem93** [reply]
@melohagan Okey, getting close now! :D 

With `{{ steps.1.rows.0.MonthlyCostTotal }}` i'm able to fetch data, but only for the first row in Services. So every column is populated with the same value. 

Is it possible to get a specific rows data, like `{{ steps.1.rows.0.Adobe.0.MonthlyCostTotal }}` ?

**Reply by melohagan** [reply]
Sounds like you are looking to do nested looping to create rows for the relationships?

In that case you would use looping to create rows for the top level.

Then you could have a second automation that has a **Row Created** trigger. Then create your related rows for that particular row that was created, i.e. the Services Monthly Cost.

**Reply by Bergem93** [reply]
@melohagan I found a "hack" which works for its purpose. By sorting the filtered rows by Service Name, i cant manually assigne data to each field by numbers.

eg. `{{ steps.1.rows.0.MonthlyCostTotal }}` will fetch Service number 1 in the db
`{{ steps.1.rows.4.MonthlyCostTotal }}` will feth the fith etc. 

Not the best solution, but it works. Thanks for all the help!
